### ISSUE TITLE
UI: On commit - empty metadata keys are allowed

### ISSUE THREAD

**Issue by N-o-Z** [treeverse/lakeFS#5251]
We should ignore metadata keys which are empty

This is how it looks in lakectl:

```
± % lakectl log lakefs://test/test3 -c ~/.lakectl-local.yaml --amount 1                                                                                                                             !10339

ID:            0f57d629f4bb2e4e5f24290a7a1c48b7249a3d11fe7b3785982f03dd6662a1a6
Author:        admin
Date:          2023-02-15 17:51:58 +0200 IST

	

Metadata:
	
	 = 
	
	test = hello
```

**Comment by zubron** [comment]
I've just put together a draft PR (#9650) and I have some thoughts and questions about metadata validation.

While implementing the UI filtering for empty metadata fields (to match CLI behavior), I noticed that we're handling this validation on the client side in both the CLI and UI rather than in the API itself.

Currently:
- CLI filters empty key-value pairs before sending to API
- UI will now filter empty key-value pairs before sending to API
- API appears to accept metadata without validation

Should metadata validation be handled by the API instead?

Having validation logic duplicated across clients seems fragile. It means that any new clients or integrations in the future will need to implement the same filtering logic if we want to have consistency. It also means the API will silently accept this form of metadata if called directly. Also, I noticed that due to the lack of validation, you can provide duplicate keys, but since metadata is stored as a map/object, one value will silently overwrite the other. Users might not realize that they aren't storing the data they thought they were (e.g. what if it was a typo that caused the duplicate key?)

Some questions:
1. Are there intentional use cases for empty keys or values in metadata?
2. Should the API validate and reject invalid metadata (empty fields, duplicate keys)?
3. Are there other metadata validation rules we should consider (e.g., key format, reserved keys, value types, limits on number of entries or size of metadata payloads)?

Moving this validation to the API would ensure consistency across all clients and provide better error messages to users. I doubt this is high priority, but I wanted to raise the question since there might be backwards compatibility concerns or design reasons for the current approach.

**Comment by N-o-Z** [comment]
> I've just put together a draft PR ([#9650](https://github.com/treeverse/lakeFS/pull/9650)) and I have some thoughts and questions about metadata validation.
> 
> While implementing the UI filtering for empty metadata fields (to match CLI behavior), I noticed that we're handling this validation on the client side in both the CLI and UI rather than in the API itself.
> 
> Currently:
> 
> * CLI filters empty key-value pairs before sending to API
> * UI will now filter empty key-value pairs before sending to API
> * API appears to accept metadata without validation
> 
> Should metadata validation be handled by the API instead?
> 
> Having validation logic duplicated across clients seems fragile. It means that any new clients or integrations in the future will need to implement the same filtering logic if we want to have consistency. It also means the API will silently accept this form of metadata if called directly. Also, I noticed that due to the lack of validation, you can provide duplicate keys, but since metadata is stored as a map/object, one value will silently overwrite the other. Users might not realize that they aren't storing the data they thought they were (e.g. what if it was a typo that caused the duplicate key?)
> 
> Some questions:
> 
> 1. Are there intentional use cases for empty keys or values in metadata?
> 2. Should the API validate and reject invalid metadata (empty fields, duplicate keys)?
> 3. Are there other metadata validation rules we should consider (e.g., key format, reserved keys, value types, limits on number of entries or size of metadata payloads)?
> 
> Moving this validation to the API would ensure consistency across all clients and provide better error messages to users. I doubt this is high priority, but I wanted to raise the question since there might be backwards compatibility concerns or design reasons for the current approach.

@zubron these are all excellent questions and it's a very good opportunity to bring @treeverse/product in the discussion.
IMO I believe there might be a use case for empty values, however we should definitely block empty keys.
Regarding where to put the validation - you made a good point regarding consistency over all clients when the server handles the validation however also consider the back and forth that could be spared if the validation is done on the client side.
Adding it on both client and server side might be a good mitigation. However if I had to choose only one I'd probably add it to the server side

**Comment by arielshaqed** [comment]
This issue makes sense if user metadata in S3 does not allow empty keys. We need to check this first. Otherwise we break the S3 gateway. 

Assuming nonempty keys are on a par with S3, I agree >100% with both @zubron and @N-o-Z. This validation MUST occur on the server side. Otherwise you can create data via the API that the UI prevents. 
Afterwards, we CAN prevent this data on the UI - I imagine an invalid key could grey out the text in the box.

**Comment by nopcoder** [comment]
I agree with the above and wanted to add that it also not just non empty key. for example s3 do not allow the key to start with space and more.

**Comment by zubron** [comment]
@treeverse/product I created some test cases and ran them against S3 and S3 gateway. There are a few instances where the gateway is not matching the S3 documentation: https://gist.github.com/zubron/783dc8ccdaba2108a5453086031e8785#file-s3-metadata-questions-md. I've included questions in that doc for discussion.

The linked gist includes the test code and the results.
